// pro-dj-link/lib/packets/cdj-status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../constants");
const header = Buffer.from([
    0x51, 0x73, 0x70, 0x74, 0x31, 0x57, 0x6d, 0x4a, 0x4f, 0x4c, 0x0a
]);
exports.mapCdjState = (flag) => ({
    playing: !!(flag & constants_1.CDJStateFlags.Play),
    sync: !!(flag & constants_1.CDJStateFlags.Sync),
    master: !!(flag & constants_1.CDJStateFlags.Master),
    onAir: !!(flag & constants_1.CDJStateFlags.OnAir)
});
exports.test = (buffer) => buffer.slice(0, header.length).equals(header);
exports.parse = (buffer) => {
    const device = buffer[0x21];
    const name = buffer.toString('utf8', 0x0b, 0x1e);
    const state = exports.mapCdjState(buffer[0x89]);
    const pitch = buffer.readInt32BE(0x8c);
    const bpm = buffer.readInt16BE(0x92);
    const current = (bpm * pitch) / 0x6400000;
    return {
        device,
        name,
        state,
        speed: {
            pitch,
            bpm,
            current
        }
    };
};
